### ISSUE TITLE
dev branch : Template Sprite problem

### ISSUE THREAD

**Issue by EnlNovius** [pythonarcade/arcade#919]
## Bug Report
I tried to use the template proposed by Tiled (ObjectLayer) with the dev version of arcade (2.6.0.dev1) and I have problems with the textures.
In this case, I have an ObjectLayer with two template objects: a tree (16x36) and a bollard (48x36), and a background which is the repetition of the same tile (16x12) (SpriteLayer). The visual in Tiled is as follows:

![image](https://user-images.githubusercontent.com/52698559/127066271-454d61f2-e7ec-46bc-b718-d884b19c64ea.png)

However the visual displayed by arcade with the following commands is not the same (see the code in the .zip file for more information):
```
self.tile_map = arcade.load_tilemap(map_name, TILE_SCALING)
self.scene = arcade.Scene.from_tilemap(self.tile_map)
```
![image](https://user-images.githubusercontent.com/52698559/127066050-c056675b-8037-40d7-9d4c-4cbddeffb48d.png)

Depending on how I build the file in Tiled, the texture displayed for the objects changes. I don't know what causes the change, but I wonder if the texture used for the templates is not the last texture added to the project (in Tiled). 

The code I use and the images are available here (normally you will just have to unzip and run `python debug.py`), I reused the code of an example from the dev documentation  : 

- [debug.zip](https://github.com/pythonarcade/arcade/files/6881507/debug.zip)

**Comment by Cleptomania** [comment]
I’ll take a look into this tomorrow. Templates have been tested in pytiled-parser, but I haven’t actually used them in a map with Arcade yet.

**Comment by EnlNovius** [comment]
By the way, does the file indicated in Sprite.texture.name correspond to the file from which the texture originates? If so, then this file is not the right one, either for the terminal or the tree, the file indicated here is the one of the ground (`ground.png-0-0-16-12-False-False-False-Simple` for the terminal and `ground.png-0-0-16-12-False-False-False-Simple` for the ground).

**Comment by EnlNovius** [comment]
While on the subject, making a first draw with gl.GL_LINEAR makes the following draws with gl.GL_LINEAR, even if the option is not specified for these draws.

**Comment by Cleptomania** [comment]
Alright I've found the problem. The templates do work out of the box, and your code is fine for that to work, your map however doesn't have the `tree` and `borne` tilesets listed in it. The tilesets only get loaded in as part of the map, so they have to be explicitly defined there. If you add the tilesets that the templates use to your map directly then it will work.

I will look into making pytiled-parser load in tilesets which are found in template files that do not exist within the map already.

EDIT: On second thought this isn't really possible, as this would mess up the GIDs within the tileset. The map needs to be aware of all tilesets that it could potentially use in order to properly load the images. If you try to load a tileset which the map file is not aware of, then you can't know what the firstgid of the tileset needs to be.

For example what's happening in your issue above.

Tiles have a GID(global ID), which is the ID of that tile within the context of the current map. That means a tile might have ID 1 in the tileset, but if your Tileset is given a `firstgid` attribute by the map of 30, then that tiles real ID is now 31. So what's happening is it's loading the tile IDs, but referencing tiles from the main tileset because the map isn't aware of the other tilesets.

Hopefully that makes sense.

**Comment by EnlNovius** [comment]
Ok thanks a lot for the help, I will systematically add the sprite in question in the map file =)

**Comment by EnlNovius** [comment]
From my last try, even if the tileset (.json) of the terminal and the tree are present in the map file (.json) (and so have a GID), they are not used.

The same map as the first one, but this time with the two templates present in the map file : 

* [debug.zip](https://github.com/pythonarcade/arcade/files/6891930/debug.zip)

**Comment by Cleptomania** [comment]
I've done a bit more testing and digging with this, there does appear to be an issue with pytiled-parser and how it handles these tilesets. I've come to the conclusion we need to handle the tilesets regardless of if they appear in the map file or not. This means pytiled-parser will need to handle modifying the GIDs accordingly and creating tilesets post map parse.

I know what's wrong but I'm not exactly sure what the fix is going to be yet, some re-structuring of the main map parsing is going to need done to properly support these. I'll keep it updated with here when I figure it all out, will definitely require a 1.6 release of pytiled-parser though.

Issue on pytiled-parser for reference: https://github.com/benjamin-kirkbride/pytiled_parser/issues/41

**Comment by Cleptomania** [comment]
Below is an image of it successfully running. I've got it working now, but this required some changes to the map parsing which breaks a lot of other tests and stuff. I'm going to do some cleanup and get tests on pytiled-parser fixed and push out v1.6.0 there, then I'll update to that version in dev branch of Arcade.

It now looks for all tilesets defined by object templates, and loads them in, automatically adjusting GID values as necessary based on all tilesets. This means it should work exactly the same regardless if your tileset is defined in the map or not. This also is a bit of an improvement such that if you use the same tileset for your main map as the objects, then you don't end up having duplicate tilesets loaded in as it checks if they've already been loaded beforehand.

The only downside to this, is that it requires your tilesets to have unique names, however I feel like that's a reasonable trade-off, and if you're not using object templates(or at least, don't expect to have different tilesets with the same name) then you'll be fine. If you do have tilesets with the same name loaded from your map file, they will work fine, as only the object templating is based on the tileset name. So if you are going to use object templates, best to just ensure you have unique tileset names.

I'll close this once I've got the new version pushed out and updated on Arcade.

![image](https://user-images.githubusercontent.com/1023162/127418642-3e3e643a-c004-4e6b-80ac-86b66764dc73.png)

**Comment by Cleptomania** [comment]
Dev branch of Arcade is now updated to pytiled-parser 1.5.2 which fixes this problem. I tried with both of your set of files that you posted and both work.

Note you'll have to force a re-install and make sure it downloads and installs the version 1.5.2 of pytiled-parser.

I'm gonna close this now, feel free to re-open if you have any other problems with it.
